# **Import Libraries**

[Metadata](https://drive.google.com/file/d/12DGAnGSPfjGqoo-isOA6SjXIRaDWtiel/view?usp=drive_link)

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
#

## **Import Data**
Created `data_original` as raw data and `data` as copy to be altered via data cleaning.

In [ ]:
data_original = pd.read_csv('/content/drive/MyDrive/Project:  Customer Segmentation Using Clustering/Databel - Data.csv')
data = data_original.copy()

## **Data Exploration**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6687 entries, 0 to 6686
Data columns (total 29 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        6687 non-null   object 
 1   Churn Label                        6687 non-null   object 
 2   Account Length (in months)         6687 non-null   int64  
 3   Local Calls                        6687 non-null   int64  
 4   Local Mins                         6687 non-null   float64
 5   Intl Calls                         6687 non-null   float64
 6   Intl Mins                          6687 non-null   float64
 7   Intl Active                        6687 non-null   object 
 8   Intl Plan                          6687 non-null   object 
 9   Extra International Charges        6687 non-null   float64
 10  Customer Service Calls             6687 non-null   int64  
 11  Avg Monthly GB Download            6687 non-null   int64

In [ ]:
data.head()


Customer ID Churn Label  Account Length (in months)  Local Calls  \
0   4444-BZPU          No                           1            3   
1   5676-PTZX          No                          33          179   
2   8532-ZEKQ          No                          44           82   
3   1314-SMPJ          No                          10           47   
4   2956-TXCJ          No                          62          184   

   Local Mins  Intl Calls  Intl Mins Intl Active Intl Plan  \
0         8.0         0.0        0.0          No        no   
1       431.3         0.0        0.0          No        no   
2       217.6         0.0        0.0          No       yes   
3       111.6        60.0       71.0         Yes       yes   
4       621.2       310.0      694.4         Yes       yes   

   Extra International Charges  ...  Senior  Group  \
0                          0.0  ...      No     No   
1                          0.0  ...      No     No   
2                          0.0  ...      No     No   
3                          0.0  ...      No     No   
4                          0.0  ...      No     No   

  Number of Customers in Group  Device Protection & Online Backup  \
0                            0                                 No   
1                            0                                Yes   
2                            0                                Yes   
3                            0                                 No   
4                            0                                 No   

    Contract Type Payment Method Monthly Charge  Total Charges Churn Category  \
0  Month-to-Month   Direct Debit             10             10            NaN   
1        One Year    Paper Check             21            703            NaN   
2        One Year   Direct Debit             23           1014            NaN   
3  Month-to-Month    Paper Check             17            177            NaN   
4        One Year   Direct Debit             28           1720            NaN   

  Churn Reason  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 29 columns]

## **Data Cleaning**

In [ ]:
# Convert column names to lowercased and underscored
data.columns = data.columns.str.lower().str.replace(' ', '_')

# Display modified column names
print("\nModified column names:")
print(data.columns)



Modified column names:
Index(['customer_id', 'churn_label', 'account_length_(in_months)',
       'local_calls', 'local_mins', 'intl_calls', 'intl_mins', 'intl_active',
       'intl_plan', 'extra_international_charges', 'customer_service_calls',
       'avg_monthly_gb_download', 'unlimited_data_plan', 'extra_data_charges',
       'state', 'phone_number', 'gender', 'age', 'under_30', 'senior', 'group',
       'number_of_customers_in_group', 'device_protection_&_online_backup',
       'contract_type', 'payment_method', 'monthly_charge', 'total_charges',
       'churn_category', 'churn_reason'],
      dtype='object')


- changed the format for ease of use and uniformity

In [ ]:
data['churn_label'].value_counts()

churn_label
No     4891
Yes    1796
Name: count, dtype: int64

- when churn is yes that means customer left business label
- there is a lot missing data for `churn category` and `churn reason` because these columns only apply to rows where churn label is "yes"
- but the

In [ ]:
data['churn_reason'].value_counts()


churn_reason
Competitor made better offer                 303
Competitor had better devices                297
Attitude of support person                   203
Don't know                                   123
Competitor offered more data                 110
Competitor offered higher download speeds     95
Attitude of service provider                  84
Price too high                                74
Product dissatisfaction                       73
Network reliability                           69
Long distance charges                         61
Service dissatisfaction                       60
Moved                                         44
Extra data charges                            37
Limited range of services                     35
Poor expertise of online support              30
Lack of affordable download/upload speed      28
Lack of self-service on Website               26
Poor expertise of phone support               11
Deceased                                       6
Name: count, dtype: int64

- there are 1796 customers who churned but only 1769 churn reasons recorded (missing 27 reasons why customers churned)

In [ ]:
data['churn_category'].value_counts()


churn_category
Competitor         805
Attitude           287
Dissatisfaction    286
Price              200
Other              191
Name: count, dtype: int64

## **Data Set for Customers Who Churned Only But Didn't Have a Reason Listed**

This data could be removed becasue we only want to know about customers with listed reasons.

In [ ]:
data_churned = data[(data['churn_label'] == "Yes") & (pd.isna(data['churn_category']))]
data_churned

customer_id churn_label  account_length_(in_months)  local_calls  \
40     7466-XCMA         Yes                          61          107   
228    9653-TWSS         Yes                          16           32   
239    5509-KHCT         Yes                          23          105   
318    9843-UGSQ         Yes                          29          113   
408    8079-UPTX         Yes                           9           48   
486    4312-TEIK         Yes                          10           27   
592    7983-XVUQ         Yes                          23           59   
637    6112-UTCV         Yes                          56          384   
702    5802-USCC         Yes                          58          367   
1291   2728-ESKY         Yes                          66          306   
1364   2067-XRSL         Yes                          17           27   
1402   1690-GNFE         Yes                          12           64   
1646   5683-DTIG         Yes                          52          110   
1829   2472-KMFK         Yes                           4            7   
1885   6655-LIXS         Yes                          26          182   
1921   3431-CBIS         Yes                          32          184   
2183   7455-LZFD         Yes                          51          124   
2291   1492-GPNX         Yes                          44          291   
2444   2854-HOLB         Yes                           1            3   
2451   9479-NYGO         Yes                           9           22   
2513   3070-ZNEE         Yes                          67          190   
2951   3220-NURQ         Yes                          21          121   
3013   3996-CPIX         Yes                          39           92   
3061   8594-SARH         Yes                          36          171   
6008   4795-WECG         Yes                          14          115   
6124   7818-YQPO         Yes                           7           14   
6277   5285-TFXI         Yes                          39          222   

      local_mins  intl_calls  intl_mins intl_active intl_plan  \
40         242.5         0.0        0.0          No        no   
228         98.6        64.0      180.8         Yes       yes   
239        348.6        23.0      285.2         Yes       yes   
318        205.2       116.0      232.0         Yes        no   
408        136.5        90.0       84.6         Yes        no   
486         60.9        60.0       87.0         Yes        no   
592        136.4       184.0      197.8         Yes        no   
637        563.3         0.0        0.0          No        no   
702        931.1       348.0      643.8         Yes       yes   
1291       795.4         0.0        0.0          No        no   
1364        69.8       136.0      139.4         Yes        no   
1402       181.7        48.0      106.8         Yes        no   
1646       261.0         0.0        0.0          No        no   
1829        15.5         0.0        0.0          No        no   
1885       411.2         0.0        0.0          No        no   
1921       446.5         0.0        0.0          No        no   
2183       355.9         0.0        0.0          No        no   
2291       699.5         0.0        0.0          No        no   
2444         7.0         0.0        0.0          No       yes   
2451        52.2         0.0        0.0          No        no   
2513       468.5         0.0        0.0          No        no   
2951       314.3        84.0      260.4         Yes        no   
3013       232.6       156.0      358.8         Yes        no   
3061       473.3       144.0      334.8         Yes        no   
6008       168.3         0.0        0.0          No        no   
6124        35.8         0.0        0.0          No        no   
6277       541.6       156.0      631.8         Yes       yes   

      extra_international_charges  ...  senior  group  \
40                            0.0  ...      No     No   
228                      

In [ ]:
# Drop rows where 'churn_label' is 'Yes' and 'churn_category' is NaN
data_without = data.drop(data[(data['churn_label'] == "Yes") & (pd.isna(data['churn_category']))].index)
data_without.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6660 entries, 0 to 6686
Data columns (total 29 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   customer_id                        6660 non-null   object 
 1   churn_label                        6660 non-null   object 
 2   account_length_(in_months)         6660 non-null   int64  
 3   local_calls                        6660 non-null   int64  
 4   local_mins                         6660 non-null   float64
 5   intl_calls                         6660 non-null   float64
 6   intl_mins                          6660 non-null   float64
 7   intl_active                        6660 non-null   object 
 8   intl_plan                          6660 non-null   object 
 9   extra_international_charges        6660 non-null   float64
 10  customer_service_calls             6660 non-null   int64  
 11  avg_monthly_gb_download            6660 non-null   int64  
 1

In [ ]:
data_without.describe()

account_length_(in_months)  local_calls   local_mins   intl_calls  \
count                 6660.000000  6660.000000  6660.000000  6660.000000   
mean                    32.345495   130.981532   322.822102    51.063085   
std                     24.612058   121.953827   288.791906   103.666036   
min                      1.000000     1.000000     4.000000     0.000000   
25%                      9.000000    31.000000    76.775000     0.000000   
50%                     29.000000    98.000000   250.500000     0.000000   
75%                     55.000000   199.000000   498.200000    50.000000   
max                     77.000000   918.000000  1234.200000  1120.000000   

         intl_mins  extra_international_charges  customer_service_calls  \
count  6660.000000                  6660.000000             6660.000000   
mean    130.065925                    33.674399                0.917417   
std     243.740901                    76.444649                1.413076   
min       0.000000                     0.000000                0.000000   
25%       0.000000                     0.000000                0.000000   
50%       0.000000                     0.000000                0.000000   
75%     140.000000                    16.300000                2.000000   
max    1372.500000                   585.800000                5.000000   

       avg_monthly_gb_download  extra_data_charges          age  \
count              6660.000000         6660.000000  6660.000000   
mean                  6.692793            3.384234    47.453153   
std                   7.458223           12.585934    16.962811   
min                   0.000000            0.000000    19.000000   
25%                   1.000000            0.000000    33.000000   
50%                   5.000000            0.000000    47.000000   
75%                   9.000000            0.000000    60.000000   
max                  43.000000           99.000000    85.000000   

       number_of_customers_in_group  monthly_charge  total_charges  
count                   6660.000000     6660.000000    6660.000000  
mean                       0.840090       31.033934    1083.821171  
std                        1.702921       16.298289    1127.586283  
min                        0.000000        5.000000       6.000000  
25%                        0.000000       16.000000     180.750000  
50%                        0.000000       31.000000     647.000000  
75%                        0.000000       43.000000    1733.250000  
max                        6.000000       78.000000    5574.000000

In [ ]:
data.describe()

account_length_(in_months)  local_calls   local_mins   intl_calls  \
count                 6687.000000  6687.000000  6687.000000  6687.000000   
mean                    32.337820   130.974129   322.752864    51.097524   
std                     24.595689   121.893966   288.619931   103.592369   
min                      1.000000     1.000000     4.000000     0.000000   
25%                      9.000000    31.000000    76.900000     0.000000   
50%                     29.000000    98.000000   250.500000     0.000000   
75%                     55.000000   199.000000   498.050000    52.000000   
max                     77.000000   918.000000  1234.200000  1120.000000   

         intl_mins  extra_international_charges  customer_service_calls  \
count  6687.000000                  6687.000000             6687.000000   
mean    130.070624                    33.641783                0.915657   
std     243.527828                    76.346828                1.411484   
min       0.000000                     0.000000                0.000000   
25%       0.000000                     0.000000                0.000000   
50%       0.000000                     0.000000                0.000000   
75%     140.400000                    16.400000                2.000000   
max    1372.500000                   585.800000                5.000000   

       avg_monthly_gb_download  extra_data_charges          age  \
count              6687.000000         6687.000000  6687.000000   
mean                  6.696276            3.374458    47.448632   
std                   7.454339           12.565309    16.969893   
min                   0.000000            0.000000    19.000000   
25%                   1.000000            0.000000    33.000000   
50%                   5.000000            0.000000    47.000000   
75%                   9.000000            0.000000    60.000000   
max                  43.000000           99.000000    85.000000   

       number_of_customers_in_group  monthly_charge  total_charges  
count                   6687.000000     6687.000000    6687.000000  
mean                       0.837894       31.030357    1083.755645  
std                        1.700780       16.288147    1127.074863  
min                        0.000000        5.000000       6.000000  
25%                        0.000000       16.000000     181.000000  
50%                        0.000000       31.000000     647.000000  
75%                        0.000000       43.000000    1732.500000  
max                        6.000000       78.000000    5574.000000

## **Creating Seperate Dataset for Churned v Non-Churned Customers**

In [ ]:
# Dataset for churned customers
churned_df = data[data['churn_label'] == 'Yes']

# Dataset for non-churned customers
non_churned_df = data[data['churn_label'] == 'No']


In [ ]:
churned_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1796 entries, 40 to 6686
Data columns (total 29 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   customer_id                        1796 non-null   object 
 1   churn_label                        1796 non-null   object 
 2   account_length_(in_months)         1796 non-null   int64  
 3   local_calls                        1796 non-null   int64  
 4   local_mins                         1796 non-null   float64
 5   intl_calls                         1796 non-null   float64
 6   intl_mins                          1796 non-null   float64
 7   intl_active                        1796 non-null   object 
 8   intl_plan                          1796 non-null   object 
 9   extra_international_charges        1796 non-null   float64
 10  customer_service_calls             1796 non-null   int64  
 11  avg_monthly_gb_download            1796 non-null   int64  
 

In [ ]:
churned_df_without = churned_df.drop(columns=['churn_label', 'churn_category', 'churn_reason']) #churned data without the churn related categories

In [ ]:
non_churned_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4891 entries, 0 to 6590
Data columns (total 29 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   customer_id                        4891 non-null   object 
 1   churn_label                        4891 non-null   object 
 2   account_length_(in_months)         4891 non-null   int64  
 3   local_calls                        4891 non-null   int64  
 4   local_mins                         4891 non-null   float64
 5   intl_calls                         4891 non-null   float64
 6   intl_mins                          4891 non-null   float64
 7   intl_active                        4891 non-null   object 
 8   intl_plan                          4891 non-null   object 
 9   extra_international_charges        4891 non-null   float64
 10  customer_service_calls             4891 non-null   int64  
 11  avg_monthly_gb_download            4891 non-null   int64  
 1

In [ ]:
non_churned_df_without = non_churned_df.drop(columns=['churn_label', 'churn_category', 'churn_reason']) #churned data without churn categories

- removing because not

## **Exporting Datasets**

In SQL I will perform data analysis

In [ ]:
project_folder_path = '/content/drive/MyDrive/Project:Customer Segmentation Using Clustering'
churned_df.to_csv(project_folder_path + '/churned_df.csv', index=False)
churned_df_without.to_csv(project_folder_path + '/churned_df_without.csv', index=False)
non_churned_df.to_csv(project_folder_path + '/non_churned_df.csv', index=False)
non_churned_df_without.to_csv(project_folder_path + '/non_churned_df_without.csv', index=False)